File: 151-160/152. Maximum Product Subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) {
    let result = nums[0];
    let preMax = nums[0];
    let preMin = nums[0];

    for (let i = 1; i < nums.length; i++) {
        let curMax = Math.max(nums[i]*preMax, nums[i]*preMin, nums[i]);
        let curMin = Math.min(nums[i]*preMax, nums[i]*preMin, nums[i]);

        preMax = curMax;
        preMin = curMin;

        result = Math.max(result, preMax);
    }
    
    return result;
};

File: 151-160/152. Maximum Product Subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        curMin, curMax = 1, 1

        for n in nums:
            tmp = curMax * n
            curMax = max(tmp, curMin * n, n)
            curMin = min(tmp, curMin * n, n)
            res = max(res, curMax)
        return res
